pkg/trace/agent/agent.go

```diff
@@ -189,50 +183,73 @@ func (a *Agent) Process(p *api.Payload) {
 	ss := new(writer.SampledSpans)
 	var envtraces []stats.EnvTrace
 	a.PrioritySampler.CountClientDroppedP0s(p.ClientDroppedP0s)
-	for _, t := range p.Traces {
-		if len(t) == 0 {
+
+	for i := 0; i < len(p.TracerPayload.Chunks); {
+		t := p.TracerPayload.Chunks[i]
+		if len(t.Spans) == 0 {
 			log.Debugf("Skipping received empty trace")
+			p.TracerPayload.RemoveChunk(i)
 			continue
 		}
 
-		tracen := int64(len(t))
+		tracen := int64(len(t.Spans))
 		atomic.AddInt64(&ts.SpansReceived, tracen)
-		err := normalizeTrace(p.Source, t)
+		err := normalizeTrace(p.Source, t.Spans)
 		if err != nil {
 			log.Debugf("Dropping invalid trace: %s", err)
 			atomic.AddInt64(&ts.SpansDropped, tracen)
+			p.TracerPayload.RemoveChunk(i)
 			continue
 		}
 
 		// Root span is used to carry some trace-level metadata, such as sampling rate and priority.
-		root := traceutil.GetRoot(t)
+		root := traceutil.GetRoot(t.Spans)
+		if t.Priority == int32(sampler.PriorityNone) && root.Metrics != nil {
+			if p, ok := root.Metrics["_sampling_priority_v1"]; ok {
+				t.Priority = int32(p)
+			}
+		}
+		// TODO: Can we expect origin to be found in root?
+		if t.Origin == "" && root.Meta != nil {
+			t.Origin = root.Meta["_dd.origin"]
+		}
 
 		if !a.Blacklister.Allows(root) {
 			log.Debugf("Trace rejected by ignore resources rules. root: %v", root)
 			atomic.AddInt64(&ts.TracesFiltered, 1)
 			atomic.AddInt64(&ts.SpansFiltered, tracen)
+			p.TracerPayload.RemoveChunk(i)
 			continue
 		}
 
 		if filteredByTags(root, a.conf.RequireTags, a.conf.RejectTags) {
 			log.Debugf("Trace rejected as it fails to meet tag requirements. root: %v", root)
 			atomic.AddInt64(&ts.TracesFiltered, 1)
 			atomic.AddInt64(&ts.SpansFiltered, tracen)
+			p.TracerPayload.RemoveChunk(i)
 			continue
 		}
 
 		// Extra sanitization steps of the trace.
-		for _, span := range t {
+		for _, span := range t.Spans {
 			for k, v := range a.conf.GlobalTags {
-				traceutil.SetMeta(span, k, v)
+				if k == "_dd.origin" {
+					t.Origin = v
+				} else {
+					traceutil.SetMeta(span, k, v)
+				}
 			}
 			a.obfuscator.Obfuscate(span)
 			Truncate(span)
 			if p.ClientComputedTopLevel {
 				traceutil.UpdateTracerTopLevel(span)
 			}
+
+			// Origin and Priority are set at chunk level
+			delete(span.Meta, "_dd.origin")
+			delete(span.Metrics, "_sampling_priority_v1")
```

Comment:
Right. Generally I'd ask for us to actually introduce a test to prove those numbers, but let's leave that out for now. So, what you're saying is:

* For old tracers, this saves us 44.6 - 14.08 = 30.52ns
* For new tracers, this costs us 14.08ns

I feel like this discussion should have more considerations, such as possibly doing this at decode time. Actually, a lot of optimisations can be done at decode time. 

Frankly, I'd leave this for a separate PR and discussion, because I wouldn't want this to postpone your work on this PR. What do you think?

Reply:
Okay, I'll remove the cleanup, then keep it in our technical dept.